src/lua/datetime.lua

```diff
@@ -420,6 +536,213 @@ local function datetime_tostring(self)
+                        format(o.min, math_fmod(ts, 60))
+    elseif abs_ts < SECS_PER_DAY then
+        s = qtail(s) .. ('%+d hours, %d minutes, %.9g seconds'):
+                        format(o.hour, math_fmod(o.min, 60), math_fmod(ts, 60))
+    else
+        s = qtail(s) .. ('%+d days, %d hours, %d minutes, %.9g seconds'):
+                        format(o.day, math_fmod(o.hour, 24),
+                               math_fmod(o.min, 60), math_fmod(ts, 60))
+    end
+    return s
+end
+
+local function normalize_nsec(secs, nsec)
+    if nsec < 0 or nsec >= NANOS_PER_SEC then
+        secs = secs + math_floor(nsec / NANOS_PER_SEC)
+        nsec = nsec % NANOS_PER_SEC
+    end
+    return secs, nsec
+end
+
+local function datetime_increment_by(self, direction, years, months,
+                                     seconds, nanoseconds)
+    -- operations with intervals should be done using local human dates
+    -- not UTC dates, thus normalize to local timezone before operations.
+    local dt = local_dt(self)
+    local secs, nsec = local_secs(self), self.nsec
+    local offset = self.tzoffset
+
+    local is_ym_updated = false
+
+    if years ~= 0 then
+        check_range(years, MIN_DATE_YEAR, MAX_DATE_YEAR, 'years')
+        local approx_dt = dt + direction * years * AVERAGE_DAYS_YEAR
+        check_dt(approx_dt, direction, years, 'years')
+        dt = builtin.tnt_dt_add_years(dt, direction * years, builtin.DT_SNAP)
+        is_ym_updated = true
+    end
+    if months ~= 0 then
+        local new_dt = dt + direction * months * AVERAGE_DAYS_MONTH
+        check_dt(new_dt, direction, months, 'months')
+        dt = builtin.tnt_dt_add_months(dt, direction * months, builtin.DT_SNAP)
+        is_ym_updated = true
+    end
+    if is_ym_updated then
+        secs = (dt - DAYS_EPOCH_OFFSET) * SECS_PER_DAY + secs % SECS_PER_DAY
+    end
+
+    if seconds ~= 0 then
+        local s, frac = math_modf(seconds)
+        secs = secs + direction * s
+        nsec = nsec + direction * frac * 1e9 -- FIXME - make sure it's integer
+    end
+
+    if nanoseconds ~= 0 then
+        nsec = nsec + direction * nanoseconds
+    end
+
+    secs, self.nsec = normalize_nsec(secs, nsec)
+    check_dt(math_floor((secs + SECS_EPOCH_OFFSET) / SECS_PER_DAY), direction)
+    self.epoch = utc_secs(secs, offset)
+    return self
+end
+
+local function date_first(lhs, rhs)
+    if is_datetime(lhs) then
+        return lhs, rhs
+    else
+        return rhs, lhs
+    end
+end
+
+local function error_incompatible(name)
+    error(("datetime:%s() - incompatible type of arguments"):
+          format(name), 3)
+end
+
+--[[
+Matrix of subtraction operands eligibility and their result type
+
+|                 | datetime | interval |
++-----------------+----------+----------+
+| datetime        | interval | datetime |
+| interval        |          | interval |
+]]
+local function datetime_interval_sub(lhs, rhs)
+    check_date_interval(lhs, "operator -")
```

Comment:
As I realized after discussion with @Mons here lies a problem:
- `__sub` /`__add` methods should allow not only interval (or date), but interval-like table also.

Reply:
Here is the slightly updated code which makes both operators `+`/`-` and methods `add`/`sub` accept either interval object or simple table (with same set of attributes):

```diff
--- a/src/lua/datetime.lua
+++ b/src/lua/datetime.lua
@@ -158,6 +158,10 @@ local function is_datetime(o)
     return ffi.istype(datetime_t, o)
 end
 
+local function is_table(o)
+    return type(o) == 'table'
+end
+
 local function check_date(o, message)
     if not is_datetime(o) then
         return error(("%s: expected datetime, but received %s"):
@@ -179,8 +183,22 @@ local function check_interval(o, message)
     end
 end
 
+local function check_interval_table(o, message)
+    if not is_table(o) and not is_interval(o) then
+        return error(("%s: expected interval or table, but received %s"):
+                     format(message, type(o)), 2)
+    end
+end
+
+local function check_date_interval_table(o, message)
+    if not is_table(o) and not is_datetime(o) and not is_interval(o) then
+        return error(("%s: expected datetime, interval or table, but received %s"):
+                     format(message, type(o)), 2)
+    end
+end
+
 local function check_table(o, message)
-    if type(o) ~= 'table' then
+    if not is_table(o) then
         return error(("%s: expected table, but received %s"):
                      format(message, type(o)), 2)
     end
@@ -276,6 +294,9 @@ local function normalize_nsec(secs, nsec)
 end
 
 local function interval_decouple_args(obj)
+    if is_interval(obj) then
+        return obj.year, obj.month, obj.sec, obj.nsec
+    end
     local year = obj.year or 0
     local month = obj.month or 0
 
@@ -285,14 +306,13 @@ local function interval_decouple_args(obj)
     secs = secs + (60 * 60) * (obj.hour or 0)
     secs = secs + 60 * (obj.min or 0)
     local sec = obj.sec
-
     local nsec, usec, msec = obj.nsec, obj.usec, obj.msec
     local count_usec = bool2int(nsec ~= nil) + bool2int(usec ~= nil) +
                        bool2int(msec ~= nil)
     if count_usec > 0 then
         if count_usec > 1 then
             error('only one of nsec, usec or msecs may be defined '..
-                  'simultaneously', 2)
+                  'simultaneously', 3)
         end
         if sec ~= nil then
             local _, fraction = math_modf(sec)
@@ -321,6 +341,10 @@ local function interval_new(obj)
     return ival
 end
 
+local function interval_init(year, month, sec, nsec)
+    return ffi.new(interval_t, sec, nsec, month, year)
+end
+
 local function nyi(msg)
     error(("Not yet implemented : '%s'"):format(msg), 3)
 end
@@ -671,34 +695,38 @@ end
 --[[
 Matrix of subtraction operands eligibility and their result type
 
-|                 | datetime | interval |
-+-----------------+----------+----------+
-| datetime        | interval | datetime |
-| interval        |          | interval |
+|                 | datetime | interval | table    |
++-----------------+----------+----------+----------+
+| datetime        | interval | datetime | datetime |
+| interval        |          | interval | interval |
+| table           |          |          |          |
 ]]
 local function datetime_interval_sub(lhs, rhs)
     check_date_interval(lhs, "operator -")
-    check_date_interval(rhs, "operator -")
-    local left_t = ffi.typeof(lhs)
-    local right_t = ffi.typeof(rhs)
+    check_date_interval_table(rhs, "operator -")
+    local left_interval = is_table(lhs) or ffi.typeof(lhs) == interval_t
+    local right_interval = is_table(rhs) or ffi.typeof(rhs) == interval_t
+    local year, month, secs, nsec
 
     -- left is date, right is interval
-    if left_t == datetime_t and right_t == interval_t then
-        return datetime_increment_by(datetime_new_copy(lhs), -1, rhs.year,
-                                     rhs.month, rhs.sec, rhs.nsec)
+    if not left_interval and right_interval then
+        year, month, secs, nsec = interval_decouple_args(rhs)
+        return datetime_increment_by(datetime_new_copy(lhs), -1, year, month,
+                                     secs, nsec)
     -- left is date, right is date
-    elseif left_t == datetime_t and right_t == datetime_t then
+    elseif not left_interval and not right_interval then
         local obj = interval_new()
         obj.sec, obj.nsec = normalize_nsec(lhs.epoch - rhs.epoch,
                                            lhs.nsec - rhs.nsec)
         return obj
     -- both left and right are intervals
-    elseif left_t == interval_t and right_t == interval_t then
-        local obj = interval_new()
-        obj.year = lhs.year - rhs.year
-        obj.month = lhs.month - rhs.month
-        obj.sec, obj.nsec = normalize_nsec(lhs.sec - rhs.sec,
-                                           lhs.nsec - rhs.nsec)
+    elseif left_interval and right_interval then
+        year, month, secs, nsec = interval_decouple_args(lhs)
+        local obj = interval_init(year, month, secs, nsec)
+        year, month, secs, nsec = interval_decouple_args(rhs)
+        obj.year = obj.year - year
+        obj.month = obj.month - month
+        obj.sec, obj.nsec = normalize_nsec(obj.sec - secs, obj.nsec - nsec)
         return obj
     else
         error_incompatible("operator -")
@@ -708,31 +736,34 @@ end
 --[[
 Matrix of addition operands eligibility and their result type
 
-|                 | datetime | interval |
-+-----------------+----------+----------+
-| datetime        |          | datetime |
-| interval        | datetime | interval |
+|                 | datetime | interval | table    |
++-----------------+----------+----------+----------+
+| datetime        |          | datetime | datetime |
+| interval        | datetime | interval | interval |
+| table           |          |          |          |
 ]]
 local function datetime_interval_add(lhs, rhs)
     local lhs, rhs = date_first(lhs, rhs)
 
     check_date_interval(lhs, "operator +")
-    check_interval(rhs, "operator +")
-    local left_t = ffi.typeof(lhs)
-    local right_t = ffi.typeof(rhs)
+    check_interval_table(rhs, "operator +")
+    local left_interval = is_table(lhs) or ffi.typeof(lhs) == interval_t
+    local right_interval = is_table(rhs) or ffi.typeof(rhs) == interval_t
+    local year, month, secs, nsec
 
     -- left is date, right is interval
-    if left_t == datetime_t and right_t == interval_t then
+    if not left_interval and right_interval then
         local obj = datetime_new_copy(lhs)
-        return datetime_increment_by(obj, 1, rhs.year, rhs.month, rhs.sec,
-                                     rhs.nsec)
+        year, month, secs, nsec = interval_decouple_args(rhs)
+        return datetime_increment_by(obj, 1, year, month, secs, nsec)
     -- both left and right are intervals
-    elseif left_t == interval_t and right_t == interval_t then
-        local obj = interval_new()
-        obj.year = lhs.year + rhs.year
-        obj.month = lhs.month + rhs.month
-        obj.sec, obj.nsec = normalize_nsec(lhs.sec + rhs.sec,
-                                           lhs.nsec + rhs.nsec)
+    elseif left_interval and right_interval then
+        year, month, secs, nsec = interval_decouple_args(lhs)
+        local obj = interval_init(year, month, secs, nsec)
+        year, month, secs, nsec = interval_decouple_args(rhs)
+        obj.year = obj.year + year
+        obj.month = obj.month + month
+        obj.sec, obj.nsec = normalize_nsec(obj.sec + secs, obj.nsec + nsec)
         return obj
     else
         error_incompatible("operator +")
@@ -778,9 +809,7 @@ end
 local function datetime_shift(self, o, direction)
     assert(direction == -1 or direction == 1)
     local title = direction > 0 and "datetime.add" or "datetime.sub"
-    if type(o) ~= 'table' then
-        error(('%s - object expected'):format(title), 2)
-    end
+    check_interval_table(o, title)
 
     local year, month, secs, nsec = interval_decouple_args(o)
     return datetime_increment_by(self, direction, year, month, secs, nsec)
diff --git a/test/app-tap/datetime.test.lua b/test/app-tap/datetime.test.lua
index 066dd25ab..518807c03 100755

```
<details>
  <summary>`datetime.test.lua` diff collapsed</summary>
  
```diff
--- a/test/app-tap/datetime.test.lua
+++ b/test/app-tap/datetime.test.lua
@@ -28,10 +28,8 @@ local timestamp_and_ymd = 'timestamp is not allowed if year/month/day provided'
 local timestamp_and_hms = 'timestamp is not allowed if hour/min/sec provided'
 local str_or_num_exp = 'tzoffset: string or number expected, but received'
 local numeric_exp = 'numeric value expected, but received '
-local add_object_expected = ('%s - object expected'):format('datetime.add')
-local sub_object_expected = ('%s - object expected'):format('datetime.sub')
-local expected_interval_but = 'expected interval, but received'
-local expected_datetime_but = 'expected datetime or interval, but received'
+local expected_interval_but = 'expected interval or table, but received'
+local expected_datetime_but = 'expected datetime, interval or table, but received'
 
 -- various error message generators
 local function exp_datetime(name, value)
@@ -643,7 +641,7 @@ test:test("Time interval __index fields", function(test)
 end)
 
 test:test("Time interval operations", function(test)
-    test:plan(66)
+    test:plan(114)
 
     local ival_new = date.interval.new
 
@@ -657,7 +655,7 @@ test:test("Time interval operations", function(test)
         {{year = -1}, '1974-08-01T00:00:00Z', '1969-10-02T00:00:00Z'},
         {{year = -1}, '1973-08-01T00:00:00Z', '1970-10-02T00:00:00Z'},
     }
-    -- check :add{}
+    -- check :add{table}
     for _, row in pairs(leap_ivals_check) do
         local delta, sadd, ssub = unpack(row)
         test:is(tostring(tadd:add(delta)), sadd,
@@ -665,15 +663,35 @@ test:test("Time interval operations", function(test)
         test:is(tostring(tsub:sub(delta)), ssub,
                 ('sub delta to %s'):format(ssub))
     end
+    -- check :add{interval}
+    tadd = date.new(base_leap_date)
+    tsub = date.new(base_leap_date)
+    for _, row in pairs(leap_ivals_check) do
+        local delta, sadd, ssub = unpack(row)
+        test:is(tostring(tadd:add(ival_new(delta))), sadd,
+                ('add ival delta to %s'):format(sadd))
+        test:is(tostring(tsub:sub(ival_new(delta))), ssub,
+                ('sub ival delta to %s'):format(ssub))
+    end
     -- check t + interval
     tadd = date.new(base_leap_date)
     tsub = date.new(base_leap_date)
     for _, row in pairs(leap_ivals_check) do
         local delta, sadd, ssub = unpack(row)
         tadd = tadd + ival_new(delta)
-        test:is(tostring(tadd), sadd, ('%s: t + delta'):format(sadd))
+        test:is(tostring(tadd), sadd, ('%s: t + ival delta'):format(sadd))
         tsub = tsub - ival_new(delta)
-        test:is(tostring(tsub), ssub, ('%s: t - delta'):format(ssub))
+        test:is(tostring(tsub), ssub, ('%s: t - ival delta'):format(ssub))
+    end
+    -- check t + {table}
+    tadd = date.new(base_leap_date)
+    tsub = date.new(base_leap_date)
+    for _, row in pairs(leap_ivals_check) do
+        local delta, sadd, ssub = unpack(row)
+        tadd = tadd + delta
+        test:is(tostring(tadd), sadd, ('%s: t + table delta'):format(sadd))
+        tsub = tsub - delta
+        test:is(tostring(tsub), ssub, ('%s: t - table delta'):format(ssub))
     end
 
     -- check average, not leap dates
@@ -692,7 +710,7 @@ test:test("Time interval operations", function(test)
          '1968-08-13T23:27:05.876544Z'},
     }
 
-    -- check :add
+    -- check :add{table}
     tadd = date.new(base_non_leap_date)
     tsub = date.new(base_non_leap_date)
     for _, row in pairs(non_leap_ivals_check) do
@@ -703,6 +721,17 @@ test:test("Time interval operations", function(test)
                 ('sub delta to %s'):format(ssub))
     end
 
+    -- check :add{interval}
+    tadd = date.new(base_non_leap_date)
+    tsub = date.new(base_non_leap_date)
+    for _, row in pairs(non_leap_ivals_check) do
+        local delta, sadd, ssub = unpack(row)
+        test:is(tostring(tadd:add(ival_new(delta))), sadd,
+                ('add delta to %s'):format(sadd))
+        test:is(tostring(tsub:sub(ival_new(delta))), ssub,
+                ('sub delta to %s'):format(ssub))
+    end
+
     -- check t + interval
     tadd = date.new(base_non_leap_date)
     tsub = date.new(base_non_leap_date)
@@ -713,6 +742,18 @@ test:test("Time interval operations", function(test)
         tsub = tsub - ival_new(delta)
         test:is(tostring(tsub), ssub, ('%s: t - delta'):format(ssub))
     end
+
+    -- check t + {table}
+    tadd = date.new(base_non_leap_date)
+    tsub = date.new(base_non_leap_date)
+    for _, row in pairs(non_leap_ivals_check) do
+        local delta, sadd, ssub = unpack(row)
+        tadd = tadd + delta
+        test:is(tostring(tadd), sadd, ('%s: t + delta'):format(sadd))
+        tsub = tsub - delta
+        test:is(tostring(tsub), ssub, ('%s: t - delta'):format(ssub))
+    end
+
     local specific_errors = {
         {only_one_of, { nsec = 123456, usec = 123}},
         {only_one_of, { nsec = 123456, msec = 123}},
@@ -727,10 +768,10 @@ test:test("Time interval operations", function(test)
         assert_raises(test, err_msg, function() return tadd:add(obj) end)
         assert_raises(test, err_msg, function() return tsub:sub(obj) end)
     end
-    assert_raises(test, add_object_expected, function() tadd:add('bogus') end)
-    assert_raises(test, add_object_expected, function() tadd:add(123) end)
-    assert_raises(test, sub_object_expected, function() tsub:sub('bogus') end)
-    assert_raises(test, sub_object_expected, function() tsub:sub(123) end)
+    assert_raises_like(test, expected_interval_but, function() tadd:add('bogus') end)
+    assert_raises_like(test, expected_interval_but, function() tadd:add(123) end)
+    assert_raises_like(test, expected_interval_but, function() tsub:sub('bogus') end)
+    assert_raises_like(test, expected_interval_but, function() tsub:sub(123) end)
 end)
 
 test:test("Months intervals with last days", function(test)
@@ -778,7 +819,7 @@ Matrix of addition operands eligibility and their result type
 | interval        |  datetime | interval |
 ]]
 test:test("Matrix of allowed time and interval additions", function(test)
-    test:plan(32)
+    test:plan(31)
 
     -- check arithmetic with leap dates
     local T1970 = date.new{year = 1970, month = 1, day = 1}
@@ -813,8 +854,6 @@ test:test("Matrix of allowed time and interval additions", function(test)
 
     assert_raises_like(test, expected_interval_but,
                        function() return T1970 + 123 end)
-    assert_raises_like(test, expected_interval_but,
-                       function() return T1970 + {} end)
     assert_raises_like(test, expected_interval_but,
                        function() return T1970 + "0" end)
 
@@ -844,7 +883,7 @@ Matrix of subtraction operands eligibility and their result type
 | interval        |           | interval |
 ]]
 test:test("Matrix of allowed time and interval subtractions", function(test)
-    test:plan(30)
+    test:plan(29)
 
     -- check arithmetic with leap dates
     local T1970 = date.new{year = 1970, month = 1, day = 1}
@@ -878,8 +917,6 @@ test:test("Matrix of allowed time and interval subtractions", function(test)
 
     assert_raises_like(test, expected_datetime_but,
                        function() return T1970 - 123 end)
-    assert_raises_like(test, expected_datetime_but,
-                       function() return T1970 - {} end)
     assert_raises_like(test, expected_datetime_but,
                        function() return T1970 - "0" end)
```
</details>